---
title: "LensCleanR"
output: 
  flexdashboard::flex_dashboard:
    theme: bootstrap
    orientation: columns
    vertical_layout: fill
runtime: shiny
---

```{r global, include=FALSE}
## LensCleanR for single cell RNA-Seq analysis with monocle
## Note tibble and bioconductor don't play nicely together
## Load data in 'global' chunk so it can be shared by all users of the dashboard
library(flexdashboard)
library(shiny)
library(shinyBS)
library(monocle)
library(HSMMSingleCell)
library(knitr)
library(plyr)
library(Matrix)
library(ggplot2)
library(reshape2)
library(Biobase)
library(data.table)
library(plotly)
library(shinyjs)
useShinyjs(rmd = TRUE)
options(shiny.maxRequestSize = 100*1024^2)

#### Colors
cbs4qual = c("#A6CEE3", "#1F78B4", "#B2DF8A", "#33A02C")
cbs4dive = c("#CA0020", "#F4A582", "#92C5DE", "#0571B0")
tol21rainbow= c("#771155", "#AA4488", "#CC99BB", "#114477", "#4477AA", "#77AADD", "#117777", "#44AAAA", "#77CCCC", "#117744", "#44AA77", "#88CCAA", "#777711", "#AAAA44", "#DDDD77", "#774411", "#AA7744", "#DDAA77", "#771122", "#AA4455", "#DD7788")


#### Load example data ahead of time
data("HSMM_expr_matrix", "HSMM_gene_annotation", "HSMM_sample_sheet")
example_hsmm <- newCellDataSet(
  as.matrix(HSMM_expr_matrix), 
  new("AnnotatedDataFrame", data = HSMM_sample_sheet),
  new("AnnotatedDataFrame", data = HSMM_gene_annotation),
  lowerDetectionLimit = 0.1,
  expressionFamily = negbinomial())

## Lung
extPath <- file.path(system.file(package = "monocle"), "extdata")
load(file.path(extPath, "lung_phenotype_data.RData"))
load(file.path(extPath, "lung_exprs_data.RData"))
load(file.path(extPath, "lung_feature_data.RData"))
example_lung <- newCellDataSet(
  lung_exprs_data[ , rownames(lung_phenotype_data)], 
  new("AnnotatedDataFrame", data = lung_phenotype_data),
  new("AnnotatedDataFrame", data = lung_feature_data),
  lowerDetectionLimit = 1,
  expressionFamily = negbinomial.size())

cds <- example_hsmm

### Reactive values
saved_values <- reactiveValues()
saved_values$cds <- example_hsmm
```


Load {data-navmenu="Pre-processing"}
=====================================  

Sidebar {.sidebar data-width=250}
-------------------------------------

```{r}
tags$hr()
### Use example dataset
selectInput("dataset", "Choose dataset:",
            c("example muscle cell data" = "hsmm", 
              "example lung cell data" = "lung",
              "upload my own data (below)" = "upload")
            )
tags$hr()

### Upload user-provided data
## Exprs data
fileInput("user_exprs", "Expression data matrix",
          accept = c("text/csv",
                     "text/comma-separated-values", 
                     "text/plain",
                     ".csv")
          )

## Pheno data
fileInput("user_pdata", "Pheno data",
          accept = c("text/csv",
                     "text/comma-separated-values", 
                     "text/plain",
                     ".csv")
          )

## Feature data
fileInput("user_fdata", "Feature data",
          accept = c("text/csv",
                     "text/comma-separated-values", 
                     "text/plain",
                     ".csv")
          )

#### User-uploaded data settings:
## Minimum expression
numericInput("lowerDetectionLimit", "Detection Limit", value = 0.1, min = 0)

## Type of expression data
selectInput("exprType", "Expression values are:",
            c("RPC (recommended)" = "rpc", "RPC sparse matrix" = "rpcSparse",
              "FPKM" = "fpkm", "Log-transformed FPKM" = "fpkmLog",
              "Raw counts" = "countsRaw")
            )
selectInput("fpkmToRPC", "Convert FPKM to RPC (recommended)", c("TRUE" = "TRUE", "FALSE" = "FALSE"))
tags$hr()

### Upload user data
actionButton("go_load_data", "Load data")
tags$hr()
```

```{r}
#### Load user-provided files
## Expression matrix
reactive({
  if (is.null(input$user_exprs)) {
    return(NULL)
    }
  saved_values$user_exprs <- read.csv(input$user_exprs$datapath, stringsAsFactors = FALSE)
  })

## Pheno data
reactive({
  if (is.null(input$user_pdata)) {
    return(NULL)
    }
  saved_values$user_pdata <- read.csv(input$user_pdata$datapath, stringsAsFactors = FALSE)
  })

## Feature data
reactive({
  if (is.null(input$user_fdata)) {
    return(NULL)
    }
  saved_values$user_fdata <- read.csv(input$user_fdata$datapath, stringsAsFactors = FALSE)
  })

### Load data
reactive({
  ### Default is example_hsmm
  dat <- example_hsmm
  
  if (input$dataset == "hsmm") {
    
    ### HSMM example dataset
    dat <- example_hsmm

  } else if (input$dataset == "lung") {
      
    ### Lung example dataset
    dat <- example_lung

  } else if (input$dataset == "upload") {
      
      ## Load user files
    user_exprs <- saved_values$user_exprs
    user_pdata <- saved_values$user_pdata
    user_fdata <- saved_values$user_fdata
    pd <- new("AnnotatedDataFrame", data = user_pdata)
    fd <- new("AnnotatedDataFrame", data = user_fdata)
    
    ## Put samples and features in order
    user_exprs <- user_exprs[order(match(rownames(user_exprs), rownames(user_fdata))), 
                order(match(colnames(user_exprs), rownames(user_pdata)))]
    
    ## Check naming
    if (rownames(user_exprs) != rownames(user_fdata)) {
      msg <- 'alert("ERROR: Feature names (e.g. gene IDs) in expression matrix (rownames) and feature data (rownames) do not match";'
      session$sendCustomMessage(type = 'jsCode', list(value = msg))
      }
    if (colnames(user_exprs) != rownames(user_pdata)) {
      msg <- 'alert("ERROR: Sample names (e.g. cell IDs) in expression matrix (colnames) and pheno data (rownames) do not match";'
      session$sendCustomMessage(type = 'jsCode', list(value = msg))
    }
    
    if (input$exprType == "countsRaw") {
      msg <- 'alert("Warning: Although Monocle can be used with raw read counts, these are not directly proportional to expression values unless you normalize them by length, so some Monocle functions could produce nonsense results. If you do not have UMI counts, We recommend you load up FPKM or TPM values instead of raw read counts.";'
      session$sendCustomMessage(type = 'jsCode', list(value = msg))
    }
    
    if (input$exprType == "fpkm") {
      cds <- newCellDataSet(as.matrix(user_exprs), pd, fd, expressionFamily = tobit())
    }
    
    if (input$exprType == "fpkmLog") {
     cds <- newCellDataSet(as.matrix(user_exprs), pd, fd, expressionFamily = gaussianff())
    }
    
    if (input$exprType == "rpcSparse") {
      cds <- newCellDataSet(as(as.matrix(user_exprs), "sparseMatrix"), pd, fd, expressionFamily = negbinomial())
    }
    
    if (input$exprType == "rpc") {
      cds <- newCellDataSet(as.matrix(user_exprs), pd, fd, expressionFamily = negbinomial())
    }
    
    if (input$fpkmToRPC == TRUE && exprType == "fpkm" | "fpkmLog") {
      rpc_matrix <- relative2abs(cds)
      cds <- newCellDataSet(as(as.matrix(rpcData), "sparseMatrix"), pd, fd, expressionFamily = negbinomial())
    }
    
    if (input$fpkmToRPC == FALSE && exprType == "fpkm" | "fpkmLog") {
      fpkmWarning <- "Note: If you performed your single-cell RNA-Seq experiment using spike-in standards, you can convert these measurements into mRNAs per cell (RPC). RPC values are often easier to analyze than FPKM or TPM values, because have better statistical tools to model them. In fact, it's possible to convert FPKM or TPM values to RPC values even if there were no spike-in standards included in the experiment"
      js_string <- 'alert("SOMETHING");'
      js_string <- sub("SOMETHING", fpkmWarning, js_string)
      session$sendCustomMessage(type = 'jsCode', list(value = js_string))
    }
    
    ## Create cds object
    dat <- newCellDataSet(
      user_exprs, 
      new("AnnotatedDataFrame", data = user_pdata),
      new("AnnotatedDataFrame", data = user_fdata),
      lowerDetectionLimit = input$lowerDetectionLimit, 
      expressionFamily = input$expressionFamily)
  }
  
  ### Save
  saved_values$cds <- dat
})

observeEvent(input$go_load_data, {
  dat <- saved_values$cds
  
  ### Estimate size factors and estimate dispersions
  dat <- estimateSizeFactors(dat)
  dat <- estimateDispersions(dat)
  
  ### Add total mRNA for each cell (to identify cells with low RNA or doublets)
  pData(dat)$mRNA <- Matrix::colSums(exprs(dat))

  ### Preview pheno Data
  output$table_pdata <- renderDataTable(
    data.frame(Biobase::pData(dat))
    )
  
  ### Preview feature Data
  if (is.null(dat$num_cells_expressed)) {
    dat <- detectGenes(dat, min_expr = 0.1)
  }
  fdata <- Biobase::fData(dat)
  output$plot_fdata <- renderPlot(
    ggplot(fdata, aes(x = num_cells_expressed, group = biotype)) + 
    geom_histogram(aes(color = biotype, fill = biotype), bins = 50, alpha = 0.1) + 
    theme_bw() +
    scale_color_manual(values = tol21rainbow[1:length(levels(fdata$biotype))]) +
    scale_fill_manual(values = tol21rainbow[1:length(levels(fdata$biotype))]) 
  )
  
  ### Save data
  saved_values$cds <- dat
})
```

Column {.tabset}
-------------------------------------

### Features

```{r}
### Error message
#textOutput("error_feature_names")

### Plot
plotOutput("plot_fdata")
```

### Pheno

```{r}
### Error message
#textOutput("error_sample_names")

### Plot
dataTableOutput("table_pdata")
```


Filter {data-navmenu="Pre-processing"}
===================================== 

```{r}
### Filter features: exclude genes expressed in few if any cells so as not to waste CPU time analyzing them for differential expression
observeEvent(input$go_filter_features, {
  dat <- saved_values$cds

  ### FIRST: Set min value for a feature to be considered expressed.
  dat <- detectGenes(dat, min_expr = input$min_expr)

  ### SECOND: Set min number of cells for feature to be expressed in.
  expressed_genes <- rownames(subset(Biobase::fData(dat), num_cells_expressed >= input$min_cells))

  ### Tabulate of # features by # cells expressing it
  table_num_cells_expr <- data.frame(table(Biobase::fData(dat)$num_cells_expr))
  table_num_cells_expr <- table_num_cells_expr[order(table_num_cells_expr$Var1, decreasing = TRUE), ]
  table_num_cells_expr$Var1 <- as.integer(table_num_cells_expr$Var1)
  rownames(table_num_cells_expr) <- NULL
  
  ### Plot
  plot_num_cells_expr <- ggplot(table_num_cells_expr, aes(x = Var1, y = Freq)) + 
      geom_bar(stat = "identity") +
      geom_vline(xintercept = input$min_cells, color = cbs4dive[1], lwd = 1.25) + 
      theme_bw() +
      labs(x = "Number of cells expressing feature", 
           y = "Features (n)")
  output$plot_num_cells_expr <- renderPlot(plot_num_cells_expr)
})


### Filter cells: restrict to singlet cells by removing doublets/triplets and cell debris
observeEvent(input$go_filter_cells, {
  dat <- saved_values$cds
  pData(dat)$mRNA <- Matrix::colSums(exprs(dat))
  
  if(is.null(input$mRNA_SD)) {
    return()
  }
  
  ### Add total mRNA for each cell (to identify cells with low RNA or doublets)
  upper <- 10^(mean(log10(pData(dat)$mRNA)) + (input$mRNA_SD*sd(log10(pData(dat)$mRNA))))
  lower <- 10^(mean(log10(pData(dat)$mRNA)) - (input$mRNA_SD*sd(log10(pData(dat)$mRNA))))

  ### Tabulate cell types 
  mRNA <- pData(dat)$mRNA
  names(mRNA) <- rownames(pData(dat))
  debris <- mRNA[mRNA < lower]
  valid_cells <- mRNA[mRNA > lower & mRNA < upper]
  doublet <- mRNA[mRNA > upper]
  table_mRNA_filtered <- cbind.data.frame(
    "Sample Category" = c("Total", "Singlets", "Doublets [filtered]", "Debris [filtered]"), 
    "Number of Cells" = rbind(
      "Total" = length(mRNA), 
      "Singlets" = length(valid_cells),
      "Doublets [filtered]" = length(double),
      "Debris [filtered]" = length(debris)))
  output$table_mRNA_filtered <- renderTable(table_mRNA_filtered)
    
  ### Add Singlet (valid_cells) column to pData, for plotting/filtering
  pData(dat)$valid_cells <- ifelse(rownames(pData(dat)) %in% names(valid_cells), TRUE, FALSE)
  saved_values$cds <- dat
  
  ### Plot cell total mRNA distribution
  pdat <- pData(dat)
  
  plot_mRNA_log <- ggplot(
    pdat, aes(x = mRNA/1E06, color = valid_cells, fill = valid_cells)) + 
    geom_density(alpha = 0.2) +
    geom_vline(xintercept = lower/1E06, lwd = 1.2) + 
    geom_vline(xintercept = upper/1E06, lwd = 1.2) +
    theme_bw() +
    scale_color_manual(values = cbs4dive[c(1,4)]) +
    scale_fill_manual(values = cbs4dive[c(1,4)]) +
    labs(x = "Total mRNA of Cell (Millions)", y = "Density",
         title = "Total mRNA Distribution (filtered)")
  output$plot_mRNA_log <- renderPlot(plot_mRNA_log)
  
  ### Filter cell samples
  dat <- dat[ , rownames(subset(pData(dat), valid_cells == TRUE))]
  dat <- detectGenes(dat, min_expr = input$min_expr)
  
  ### Save
  saved_values$cds_filt <- dat
})
```

Inputs {.sidebar}
-------------------------------------

```{r}
### Minimum expression value
sliderInput("min_expr", "Min. expression value of feature",
            min = 0, max = 10, step = 0.01, value = 0.1)

### Minimum number of cells
sliderInput("min_cells", "Min. number of cells expressing feature",
            min = 0, max = 100, step = 1, value = 10)

### Filter features
actionButton("go_filter_features", "Filter features")

### Filter: Remove doublets and debris-cells (keep singlets only)
sliderInput("mRNA_SD", "Number of SDs from mean Total mRNA for singlet cell cutoff",
             min = 0, max = 5, step = 0.25, value = 2)
### Submit
actionButton("go_filter_cells", "Filter cells")
```

Column {.tabset}
-------------------------------------

### Filter features

```{r}
### Histogram of cells expressing feature based on min_expr
plotOutput("plot_num_cells_expr")
```


### Filter cells

```{r}
### Plot
plotOutput("plot_mRNA_log")
```

### Summary

```{r}
### Table
tableOutput("table_mRNA_filtered")
```


Exploratory {data-navmenu="Pre-processing"}
===================================== 

```{r}
### Select column in table -> show details in inputs
observeEvent(input$level_xaxis, {
  dat <- saved_values$cds
  if (is.null(input$level_xaxis)) {
    level_xvar <- "select data level to filter"
  } 
  if (input$level_xaxis == "fData") {
    level_xvar <- colnames(fData(dat))
  }
  if (input$level_xaxis == "pData") {
    level_xvar <- colnames(pData(dat))
  }
  updateSelectInput(session, "level_xvar", choices = level_xvar)
})

observeEvent(input$level_yaxis, {
  dat <- saved_values$cds
  if (is.null(input$level_yaxis)) {
    level_yvar <- "select data level to filter"
  } 
  if (input$level_yaxis == "fData") {
    level_yvar <- colnames(fData(dat))
  }
  if (input$level_yaxis == "pData") {
    level_yvar <- colnames(pData(dat))
  }
  updateSelectInput(session, "level_yvar", choices = level_yvar)
})

### Plot
observeEvent(input$go_plot_exploratory, {
  dat <- Biobase::pData(saved_values$cds_filt)
  updateSelectizeInput(session, "select_genes", 
                       choices = fData(saved_values$cds_filt)$gene_short_name, 
                       server = TRUE)
  if (!is.null(input$genes)) {
    output$plot <- renderPlot(
      qplot(input$level_xvar, input$level_yvar, data = dat, color = input$genes))
  } else 
    output$plot <- renderPlot(
      qplot(input$level_xvar, input$level_yvar, data = dat))

  output$plot_clickinfo <- renderPrint({
    cat("Click:\n")
    str(input$plot_click)
    })

  output$plot_hoverinfo <- renderPrint({
    cat("Hover (viewpoints):\n")
    str(input$plot_hover)
  })

  output$plot_brushinfo <- renderPrint({
    cat("Brush (debounced):\n")
    str(input$plot_brush)
  })
})

#### THESE AREN'T WORKING!! ####


#### Warning: Error in [: object of type 'environment' is not subsettable
output$plot_clickedpoints <- renderTable({
  res <- nearPoints(saved_values$plot_dat, input$plot_click, 
                    xvar = input$level_xvar, yvar = input$level_yvar)
  if (nrow(res) == 0)
    return()
  res
})

#### ?Error
output$plot_brushedpoints <- renderTable({
  res <- brushedPoints(saved_values$plot_dat, input$plot_brush, allRows = TRUE)
  if (nrow(res) == 0)
    return()
  res  
})

#### Download brushed data as .csv
output$download_brushedpoints <- downloadHandler(
  filename = function() paste0('BrushedPoints', Sys.Date(), Sys.time(), '.csv', sep = '_'),
  content = function(file) write.csv(data(), file))
```


Inputs {.sidebar}
-------------------------------------

```{r}
### Choose data levels to plot
selectInput("level_xaxis", "Choose data level to plot on x-axis:",
            c("fData" = "fData", "pData" = "pData"))
selectInput("level_yaxis", "Choose 2nd data level to plot on y-axis:",
            c("fData" = "fData", "pData" = "pData"))

### Choose variables to plot
selectInput("level_xvar", "Choose variable to plot on x-axis:", NULL)
selectInput("level_yvar", "Choose variable to plot on y-axis:", NULL)

### Choose gene(s) to highlight
selectizeInput("select_genes", "Gene symbol", choices = NULL, options = list(placeholder = "select a gene")) 

### Action button to plot
actionButton("go_plot_exploratory", "Plot")

### Download brushed samples
downloadButton("download_brushedpoints", "Download Brushed Points")
```

Row {row-height=400}
-------------------------------------

Column {column-width=300}
-------------------------------------

```{r eval=FALSE}
### Plot
plotOutput("plot", height = 500,
           click = "plot_click",
           hover = hoverOpts(id = "plot_hover", delayType = "throttle"),
           brush = brushOpts(id = "plot_brush"))
```

Column {column-width=100}
-------------------------------------

```{r eval=FALSE}
### Plot: brush info.
wellPanel(actionButton("go_newplot", "Launch"))
verbatimTextOutput("plot_brushinfo")
```

Row {row-height=400}
-------------------------------------

Column {column-width=200}
-------------------------------------

```{r eval=FALSE}
h4("Clicked points")
tableOutput("plot_clickedpoints")
h4("Brushed points")
tableOutput("plot_brushedpoints")
```

Column {column-width=200}
-------------------------------------

```{r eval=FALSE}
### Plot: click/hover info.
verbatimTextOutput("plot_clickinfo")
verbatimTextOutput("plot_hoverinfo")
```


Cell-type hierarchy {data-navmenu="Classify"}
=====================================

```{r eval=FALSE}
## Classify and count cells
## User provides genes that mark a population of cells
  gene1_id <- reactive({
    rownames(subset(fData(cdsFilt), gene_short_name == input$gene1))
  })
  gene2_id <- reactive({
    rownames(subset(fData(cdsFilt), gene_short_name == input$gene2))
  })
  ## Need to print warning if gene name is not in fData(cds), and remind user of nomenclature
  cell1_id <- reactive({
    input$cell1
  })
  cell2_id <- reactive({
    input$cell2
  })
  
  ## Add cell type information
  cth <- newCellTypeHierarchy()
  cth <- addCellType(cth, cell1_id, classify_func = function (x) {
    x[gene1_id, ] >= 1
    })
  cth <- addCellType(cth, cell2_id, classify_func = function (x) {
    x[gene1_id, ] < 1 & x[gene2_id, ] > 1
    })
  
  ## Classify cells based on that information
  cds <- classifyCells(cds, cth, input$min_expr)
  
  ## Summarize results in table
  renderTable(table(pData(cds)$CellType))
  
  ## Summarize results with pie chart
  plot.supervised <- ggplot(pData(cds), aes(x = factor(1), fill = factor(CellType))) +
    geom_bar(width = 1) + 
    coord_polar(theta = "y") +
    theme(axis.title = element_blank())
  renderPlot(plot.supervised)
```


Unsupervised clustering {data-navmenu="Classify"}
=====================================

```{r eval=FALSE}
disp_table <- dispersionTable(cds)
  unsup_clustering_genes <- subset(disp_table, 
                                   mean_expression >= input$mean_expr & 
                                     dispersion_empirical >= 1 * dispersion_fit)
  
  ## Plot
  ## Need to add user input for what covariates/attributes to include in the model (eg Media)
  ## Need to make the rendering iterative/interactive
  cds <- setOrderingFilter(cds, unsup_clustering_genes$gene_id)
  plot_ordering_genes(cds)
  cds <- clusterCells(cds, num_clusters = input$ncell)
  plot_cell_trajectory(cds, 1, 2, color = "CellType")
  plot_cell_trajectory(cds, 1, 2, color = "Media")
  cds <- clusterCells(cds, residualModelFormulaStr = "~Media", num_clusters = input$ncell)
  plot_cell_trajectory(cds, 1, 2, color = "Media")
  cds <- clusterCells(cds, residualModelFormulaStr="~Media + num_genes_expr", num_clusters = input$ncell)
  plot_cell_trajectory(cds, 1, 2, color = "num_genes_expr")
  plot_cell_trajectory(cds, 1, 2, color = "CellType")
```
  
  
Semi-supervised clustering {data-navmenu="Classify"}
=====================================

```{r eval=FALSE}
## Semi-supervised clustering with known gene markers
marker_diff <- markerDiffTable(cds[expressed_genes, ],
                                 cth,
                                 residualModelFormulaStr = "~Media", cores = 1)
  candidate_clustering_genes <- rownames(subset(marker_diff, qval < input$qval))
  marker_spec <- calculateMarkerSpecificity(cds[candidate_clustering_genes, ], cth)
  
  ## Print top 3 gene markers
  output$topGenes <- head(selectTopMarkers(marker_spec, 3))
  
  ## Use the top 20 markers for clustering
  semisup_clustering_genes <- unique(selectTopMarkers(marker_spec, 20)$gene_id)
  
  ## Plot
  cds <- setOrderingFilter(cds, semisup_clustering_genes)
  renderPlot(plot_ordering_genes(cds))
  cds <- clusterCells(cds,
                       num_clusters = input$ncell,
                       clustering_genes = semisup_clustering_genes,
                       residualModelFormulaStr = "~Media + num_genes_expr")
  renderPlot(plot_cell_trajectory(cds, 1, 2, 
                                  color = "CellType"))
```


Impute cell-types {data-navmenu="Classify"}
=====================================

```{r eval=FALSE}
  cds <- clusterCells(cds,
                       num_clusters = input$ncell,
                       frequency_thresh = input$freq,
                       cell_type_hierarchy = cth,
                       clustering_genes = rownames(subset(marker_diff, qval < qval)),
                       residualModelFormulaStr = "~Media + num_genes_expr")
  
  ## Plot
  renderPlot(plot_cell_trajectory(cds, 1, 2, 
                                  color = "CellType", 
                                  markers = c(gene1_id, gene2_id)))
  plot.imputed <- ggplot(pData(cds), aes(x = factor(1), fill = factor(CellType))) +
    geom_bar(width = 1) + coord_polar(theta = "y") +
    theme(axis.title = element_blank())
  renderPlot(plot.imputed)
  cds1 <- cds[ , pData(cds)$CellType == cds1]
  cds1 <- estimateDispersions(cds1)
```


Unsupervised ordering {data-navmenu="Pseudotime"}
=====================================

```{r eval=FALSE}
reactive({
  dat <- saved_res$cds1
  diff_test_res <- differentialGeneTest(dat[saved_values$expressed_genes, ],
                                        fullModelFormulaStr = "~Media")
  ordering_genes <- rownames(subset(diff_test_res, qval < input$qval))
  disp_table <- dispersionTable(dat)
  ordering_genes <- subset(disp_table,
                           mean_expression >= input$mean_exp &
                             dispersion_empirical >= 2 * dispersion_fit)$gene_id
  dat <- setOrderingFilter(dat, ordering_genes)
  saved_values$cds1 <- dat
})
  
  ## Plots
observeEvent(input$go_plot_2.1{
  dat <- saved_res$cds1
  output$plot_2.1.a <- renderPlot(plot_ordering_genes(dat))
  dat <- reduceDimension(dat, 
                         #max_components = input$max_components,
                         max_components = 2)
  dat <- orderCells(dat, reverse = FALSE)
  
  ## Plot all genes
  output$plot_2.1.b <- renderPlot(plot_cell_trajectory(dat, color_by = "Hours"))
  
  ## Plot subset of genes
  dat_expr_genes <- rownames(subset(fData(dat), num_cells_expressed >= input$num_cells_expressed))
  datFilt <- dat[dat_expr_genes,]
  my_genes <- rownames(subset(fData(cds1Filt), gene_short_name %in% input$genesUser))
  dat_subset <- datFilt[my_genes, ]
  output$plot_2.1.c <- renderPlot(plot_genes_in_pseudotime(dat_subset, color_by = "Hours"))
  
  saved_values$cds1Filt <- datFilt
})
```

Ordering with PCA-selected genes {data-navmenu="Pseudotime"}
=====================================

```{r eval=FALSE}
reactive({
  dat <- saved_values$cds1Filt
  expr_Filt <- t(t(exprs(dat)/pData(dat)$Size_Factor))
  nz_genes <- which(expr_Filt != 0)
  expr_Filt[nz_genes] <- log(expr_Filt[nz_genes] + 1)
  
  ## Calculate variance across genes without converting to dense matrix
  expr_means <- Matrix::rowMeans(expr_Filt)
  expr_vars <- Matrix::rowMeans((expr_Filt - expression_means)^2)
  
  ## Filter out genes that are constant across all cells
  genes_to_keep <- expr_vars > 0
  expr_filt <- expr_Filt[genes_to_keep, ]
  expr_means <- expr_means[genes_to_keep]
  expr_vars <- expr_vars[genes_to_keep]
  
  ## Take top PCA loading genes
  irlba_pca_res <- irlba(t(expr_Filt),
                         nu = 0,
                         center = expr_means,
                         scale = sqrt(expr_vars),
                         right_only = TRUE)$v
  rownames(irlba_pca_res) <- rownames(expr_Filt)
  PC2_genes <- names(sort(abs(irlba_pca_res[ , 2]), decreasing = TRUE))[1:100]
  PC3_genes <- names(sort(abs(irlba_pca_res[ , 3]), decreasing = TRUE))[1:100]
  ordering_genes <- union(PC2_genes, PC3_genes)
  dat <- setOrderingFilter(dat, ordering_genes)
  dat <- reduceDimension(dat, max_components = 2)
  dat <- orderCells(dat, reverse = FALSE)
  saved_values$cds1Filt <- dat
})
  
  ## Plot
observeEvent({
  output$plot_2.2 <- renderPlot(plot_cell_trajectory(saved_values$cds1Filt, color_by = "Hours"))
})
```

Semi-supervised ordering {data-navmenu="Pseudotime"}
=====================================

```{r eval=FALSE}
### Semi-supervised ordering with known marker genes
reactive({
  dat <- saved_values$cds1
  cth <- newCellTypeHierarchy()
  cth <- addCellType(cth, cell1_id, classify_func = function (x) {
    x[gene1_id, ] >= 1
    })
  cth <- addCellType(cth, cell2_id, classify_func = function (x) {
    x[gene2_id, ] >= 1
    })
  marker_diff <- markerDiffTable(dat[saved_values$expressed_genes, ],
                                 cth,
                                 cores = 1)
  semisup_clustering_genes <- rownames(subset(marker_diff, qval < input$qval))
  
  ## Print number
  renderTable(length(semisup_clustering_genes))
  
  cds1 <- setOrderingFilter(cds1, semisup_clustering_genes)
  renderPlot(plot_ordering_genes(cds1))
  cds1 <- reduceDimension(cds1, max_components = 2)
  cds1 <- orderCells(cds1, reverse = TRUE)
  
  ## Plot
  renderPlot(plot_cell_trajectory(cds1, color_by = "Hours"))
  
  ## Plot subset of genes
  cds1Filt <- cds1[saved_values$expressed_genes, ]
  my_genes <- rownames(subset(fData(cds1Filt), gene_short_name %in% input$genesUser))
  cds_subset <- cds1Filt[my_genes, ]
  renderPlot(plot_genes_in_pseudotime(cds_subset, color_by = "Hours"))
})
  
## Reconstructing branched trajectories Plot
reactive({
  dat <- saved_values$cds_filt
  plot.traject <- plot_cell_trajectory(orderCells(dat, root_state = input$root_state),
                                       color_by = input$color_by)
  renderPlot(plot.traject)
  #renderPlot(plot.traject + facet_wrap(~input$facet_by))
})
```
  
Branches {data-navmenu="Pseudotime"}
=====================================

```{r eval=FALSE}
dat <- saved_values$cds_filt
  
renderPlot(plot_cell_trajectory(
  dat, 
  color_by = "Time"))
BEAM_res <- BEAM(dat, branch_point = 1, cores = 1)
BEAM_res <- BEAM_res[order(BEAM_res$qval), ]
BEAM_res <- BEAM_res[ , c("gene_short_name", "pval", "qval")]
  
## Plot heatmap
renderPlot(plot_genes_branched_heatmap(
  dat[rownames(subset(BEAM_res, qval < qval)), ],
  branch_point = 1,
  num_clusters = 4,
  cores = 1,
  use_gene_short_name = TRUE,
  show_rownames = TRUE))
dat_genes <- rownames(subset(fData(dat), gene_short_name %in% genesUser))
  
## Plot pseudotime
renderPlot(plot_genes_branched_pseudotime(
  dat[dat_genes, ],
  branch_point = 1,
  color_by = "Time",
  ncol = 1))
```

Basic {data-navmenu="Differential expression"}
=====================================

```{r eval=FALSE}
  marker_genes <- rownames(subset(fData(cds), gene_short_name %in% genesUser))
  diff_test_res <- differentialGeneTest(cds1Filt[marker_genes, ],
                                        fullModelFormulaStr = "~Media")
  sig_genes <- subset(diff_test_res, qval < qval)
  
  ## Print results
  renderTable(sig_genes[ , c("gene_short_name", "pval", "qval")])
  
  ## Plot
  cds_subset <- cds1Filt[rownames(subset(fData(cds1Filt), gene_short_name %in% genesUser)), ]
  renderPlot(plot_genes_jitter(cds_subset, grouping = "Media", ncol = 2))

  #### Finding genes that distinguish cell type or state
  to_be_tested <- rownames(subset(fData(cds), gene_short_name %in% genesUser))
  cds_subset <- cds[to_be_tested, ]
  diff_test_res <- differentialGeneTest(cds_subset, fullModelFormulaStr = "~CellType")
  
  ## Print results
  renderTable(diff_test_res[ , c("gene_short_name", "pval", "qval")])
  
  ## Plot
  renderPlot(plot_genes_jitter(cds_subset, 
                               grouping = "CellType", 
                               color_by = "CellType",
                               nrow = 1, 
                               ncol = NULL, 
                               plot_trend = TRUE))
  full_model_fits <- fitModel(cds_subset, modelFormulaStr = "~CellType")
  reduced_model_fits <- fitModel(cds_subset, modelFormulaStr = "~1")
  diff_test_res <- compareModels(full_model_fits, reduced_model_fits)
  
  ## Print results
  renderTable(diff_test_res)

  #### Finding genes that change as a function of pseudotime
  to_be_tested <- rownames(subset(fData(cds), gene_short_name %in% genesUser))
  cds_subset <- cds1[to_be_tested, ]
  diff_test_res <- differentialGeneTest(cds_subset, fullModelFormulaStr = "~sm.ns(Pseudotime)")
  
  ## Print results
  renderTable(diff_test_res[ , c("gene_short_name", "pval", "qval")])
  
  ## Plot
  renderPlot(plot_genes_in_pseudotime(cds_subset, color_by = "Hours"))
  
  ## Clustering genes by pseudotemporal expression pattern
  diff_test_res <- differentialGeneTest(cds1[marker_genes, ],
                                        fullModelFormulaStr = "~sm.ns(Pseudotime)")
  sig_gene_names <- rownames(subset(diff_test_res, qval < qval))
  
  ## Plot
  plot.pseudoCluster <- plot_pseudotime_heatmap(cds1[sig_gene_names, ], 
                                                num_clusters = 2, 
                                                cores = 1, 
                                                show_rownames = TRUE)
  renderPlot(plot.pseudoCluster)
```
  
  
Multi-factorial {data-navmenu="Differential expression"}
=====================================

```{r eval=FALSE}
  to_be_tested <- rownames(subset(fData(cds), gene_short_name %in% genesUser))
  cds_subset <- cds[to_be_tested, ]
  diff_test_res <- differentialGeneTest(cds_subset,
                                        fullModelFormulaStr = "~CellType + Hours",
                                        reducedModelFormulaStr = "~Hours")
  diff_test_res[ , c("gene_short_name", "pval", "qval")]
  renderPlot(plot_genes_jitter(cds_subset,
                               grouping = "Hours", 
                               color_by = "CellType", 
                               plot_trend = TRUE) +
               acet_wrap(~feature_label, scales = "free_y"))
```


About
=====================================

### About the LensCleanR shiny app

```{r}
renderPrint({
  "Based on the monocle v2 tutorial.";
  "Developed by (in alphabetical order) Gui G, Mohammadyani D, Norris AL, and Workman RE, under the guidance of Jeff Leek, Sean Kross, and Loyal Goff as part of the Data Apps Course at Johns Hopkins School of Medicine in Spring 2017."
})
```
